go/datastore/internal/client/postgres/client.go

```diff
@@ -0,0 +1,153 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"net/url"
+	"strings"
+
+	_ "github.com/jackc/pgx/v5/stdlib"
+	"github.com/pluralsh/console/go/datastore/api/v1alpha1"
+	"github.com/pluralsh/console/go/datastore/internal/utils"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type client struct {
+	ctx        context.Context
+	connection string
+}
+
+type Client interface {
+	Init(ctx context.Context, client k8sclient.Client, credentials *v1alpha1.PostgresCredentials) error
+	Ping() error
+	DeleteDatabase(dbName string) error
+	UpsertDatabase(dbName string) error
+}
+
+func New() Client {
+	return &client{}
+}
+
+func (c *client) Init(ctx context.Context, client k8sclient.Client, credentials *v1alpha1.PostgresCredentials) error {
+	secret, err := utils.GetSecret(ctx, client, &corev1.SecretReference{Name: credentials.Spec.PasswordSecretKeyRef.Name, Namespace: credentials.Namespace})
+	if err != nil {
+		return err
+	}
+
+	key, exists := secret.Data[credentials.Spec.PasswordSecretKeyRef.Key]
+	if !exists {
+		return fmt.Errorf("secret %s does not contain key %s", credentials.Spec.PasswordSecretKeyRef.Name, credentials.Spec.PasswordSecretKeyRef.Key)
+	}
+
+	password := strings.ReplaceAll(string(key), "\n", "")
+
+	u := &url.URL{
+		Scheme: "postgres",
+		User:   url.UserPassword(credentials.Spec.Username, password),
+		Host:   fmt.Sprintf("%s:%d", credentials.Spec.Host, credentials.Spec.Port),
+		Path:   "postgres",
```

Comment:
I don't think we should assume login to the postgres path, it could be in the credentials too (as database in the spec)